### ISSUE TITLE
NaN values with CategoricalEncoder if rare labels only in test

### ISSUE THREAD

**Issue by ftvalentini** [solegalli/feature_engine#105]
It seems that categorical encoders (for example MeanCategoricalEncoder) cannot encode rare unseen labels in test data even after using RareLabelEncoder. This happens when there are rare labels only in test **but not in train**. 
This result is somehow reasonable because there is no way to impute a value to a "Rare" category if there were no "Rare" during training -- what's the workaround to avoid NaN in test?

```python 
import pandas as pd
from sklearn.pipeline import Pipeline
from feature_engine import categorical_encoders as ce

X_train = pd.DataFrame({'var': ['A','A','B','B']})
y_train = pd.DataFrame({'y': [1,2,3,4]})
X_test = pd.DataFrame({'var': ['A','A','C']})
pipe = Pipeline([
   ('rare_encoder', ce.RareLabelCategoricalEncoder(tol=0.01, n_categories=1))
    ,('mean_encoder', ce.MeanCategoricalEncoder())
])
pipe.fit_transform(X_train, y_train)
pipe.transform(X_test)
```

The output:

```
   var
0  1.5
1  1.5
2  NaN
/.../feature_engine/base_transformers.py:91: UserWarning: NaN values were introduced in the returned dataframe by the encoder.This means that some of the categories in the input dataframe were not present in the training set used when the fit method was called. Thus, mappings for those categories does not exist. Try using the RareLabelCategoricalEncoder to remove infrequent categories before calling this encoder.
  "NaN values were introduced in the returned dataframe by the encoder."
```

**Comment by solegalli** [comment]
Hello @ftvalentini 

Technically this is possible:
Set  up the new parameter of the RareLabelCategoricalEncoder `max_n_categories` to 1. That is:

`ce.RareLabelCategoricalEncoder(tol=0.01, n_categories=1, max_n_categories=1)`

docs: https://feature-engine.readthedocs.io/en/latest/encoders/RareLabelCategoricalEncoder.html

That should return something like:
X_train['var']=['A', 'A', 'Rare', 'Rare']
X_test['var']=['A','A','Rare']

In short, when this happens, we kind of need to ensure that there is a 'Rare' category in the training set either by increasing the tolerance, or reducing the max_n_categories or n_categories.

Alternatively, the one hot encoder into k binary variables can work as well. If the category did not appear in the dataset, it will show 0 in all the binaries.

Cheers
Sole
